### ISSUE TITLE
libdrm doesn't build against musl (regression)

### ISSUE THREAD

**Issue by nh2** [NixOS/nixpkgs#66441]
I'm in the process of upgrading `static-haskell-nix` to latest nixpkgs `master` (commit 07ce48cc3a1c6551a754ad487d2d9a90361530e5), and get the following build failure for `libdrm`:

```
[58/87] Compiling C object 'tests/nouveau/e47a46e@@threaded@exe/threaded.c.o'.
FAILED: tests/nouveau/e47a46e@@threaded@exe/threaded.c.o 
/nix/store/ad3j1a9myvk162458qpj8qzw14kdmm2w-gcc-wrapper-7.4.0/bin/cc -Itests/nouveau/e47a46e@@threaded@exe -Itests/nouveau -I../tests/nouveau -I. -I../ -I../include/drm -Itests/nouveau>
../tests/nouveau/threaded.c:39:5: error: conflicting types for ‘ioctl’
 int ioctl(int fd, unsigned long request, ...)
     ^~~~~
In file included from ../tests/nouveau/threaded.c:23:0:
/nix/store/dd1il3gf9q3386pk68kik56mn52pvs7j-musl-1.1.22-dev/include/sys/ioctl.h:119:5: note: previous declaration of ‘ioctl’ was here
 int ioctl (int, int, ...);
     ^~~~~
ninja: build stopped: subcommand failed.
```

The problem didn't occur before commit 4c77d471e40ae1b1581cdbc33fb8fae00e6fd0b7 `libdrm: build with meson` (from #57773) because the tests weren't built by default with autotools. See
https://github.com/openembedded/openembedded-core/commit/8dcebad54ded4fa9b3455b674be37ad0b10190d8#diff-4295180ffa50d118533fe15edfd3c729 which says

> As Meson always builds the test suite (instead of building before running it),
add a patch to fix compilation with musl.

The compile problem seems to be that glibc decided to not implement the POSIX type of `ioctl()`, but to change the second parameter from `int` to `unsigned long`).

Most projects `#ifdef` that problem:

* https://bugs.freedesktop.org/show_bug.cgi?id=85319
* https://patchwork.linuxtv.org/patch/28041/

Other distros carry patches for `libdrm` because of that:

* OpenEmbedded: https://github.com/openembedded/openembedded-core/blob/master/meta/recipes-graphics/drm/libdrm/musl-ioctl.patch
* Alpine Linux: https://git.alpinelinux.org/aports/commit/main/libdrm/ioctl.patch?id=7ba3523fe098c050e2729a4cd9a1fd50e2167aca
* Gentoo: https://gitweb.gentoo.org/proj/musl.git/diff/x11-libs/libdrm/files/fix-ioctl.patch?id2=5600c0735d024b65c378a86f36a9d128d2111fb6

Now that we use meson, we should most likely also carry such a patch.

CC @colemickens @dtzWill

**Comment by matthewbauer** [comment]
@nh2 I'm still seeing this. Does b577340eb5bc3b72549f0544b50e2e37df78bf12 fix the issue? If so, we can apply it to master.

**Comment by nh2** [comment]
> @nh2 I'm still seeing this. Does [b577340](https://github.com/NixOS/nixpkgs/commit/b577340eb5bc3b72549f0544b50e2e37df78bf12) fix the issue? If so, we can apply it to master.

Yes, it fixes it, and we should in your #69270.
